tachyon/crypto/sumcheck/sumcheck_prover.h

```diff
@@ -0,0 +1,180 @@
+//
+// The polynomial is represented by a list of terms each containing a
+// coefficient and a list of |MultilinearDenseEvaluations|. Refer to
+// "tachyon/math/polynomials/multivariate/linear_combination.h" for more info.
+template <typename MLE, size_t MaxDegree>
+class SumcheckProver {
+ public:
+  constexpr static size_t kParallelFactor = 16;
+
+  using F = typename MLE::Field;
+  using Point = typename MLE::Point;
+  using Term = math::LinearCombinationTerm<F>;
+
+  SumcheckProver() = default;
+  explicit SumcheckProver(const SumcheckProvingKey<MLE>& key)
+      : num_variables_(key.verifying_key.num_variables),
+        max_evaluations_(key.verifying_key.max_evaluations),
+        terms_(key.terms),
+        flattened_ml_evaluations_(key.flattened_ml_evaluations) {
+    // Constants are not accepted
+    CHECK_NE(num_variables_, size_t{0});
+    randomness_.reserve(num_variables_);
+  }
+  explicit SumcheckProver(SumcheckProvingKey<MLE>&& key)
+      : num_variables_(key.verifying_key.num_variables),
+        max_evaluations_(key.verifying_key.max_evaluations),
+        terms_(std::move(key.terms)),
+        flattened_ml_evaluations_(std::move(key.flattened_ml_evaluations)) {
+    // Constants are not accepted
+    CHECK_NE(num_variables_, size_t{0});
+    randomness_.reserve(num_variables_);
+  }
+
+  // Generate prover message, and proceed to next round.
+  //
+  // Main algorithm used is from section 3.2 of
+  // [XZZPS19](https://eprint.iacr.org/2019/317.pdf#subsection.3.2).
+  //
+  // Referencing https://people.cs.georgetown.edu/jthaler/sumcheck.pdf
+  // For v = the number of variables...
+  // For the first round, the Prover creates and sends:
+  //   g₁(X₁) := ∑ g(X₁,x₂,...,xᵥ), (x₂,...,xᵥ)∈{0,1}ᵛ⁻¹
+  // For the consequent round i, the Prover creates and sends:
+  //   gᵢ(Xᵢ) = ∑ g(r₁,...,rᵢ₋₁,Xᵢ,xᵢ₊₁,...,xᵥ), (xᵢ₊₁ ,...,xv)∈{0,1}ᵛ⁻¹
+  // More specifically, |Prover(VerifierMsg)| fixes the front variables with
+  // r₁,...,rᵢ₋₁, and |Prover()| creates the total gᵢ(Xᵢ).
+  SumcheckProverMsg<F, MaxDegree> Round() {
+    // Max number of prover rounds should be |num_variables_|.
+    CHECK_LE(++round_, num_variables_);
+
+    // Generate sum
+    // clang-format off
+    // g(x₁, x₂) = 1(1 - x₁)(1 - x₂) + 2x₁(1 - x₂) + 3(1 - x₁)x₂ + 4x₁x₂
+    //
+    // Creating the univariate polynomial:
+    // g₁(X₁) = ∑ 1(1 - X₁)(1 - x₂) + 2X₁(1 - x₂) + 3(1 - X₁)x₂ + 4X₁x₂, x₂∈{0,1}
+    //        = 1(1 - X₁) + 2X₁ + 3(1 - X₁) + 4X₁
+    //        = start₀ + step₀X₁ + start₁ + step₁X₁
+    //         (where start₀ = 1, step₀ = 1 = 2 - 1, start₁ = 3 and step₁ = 1 = 4 - 3)
+    //        = start₀ + start₁ + (step₀ + step₁)X₁
+    //
+    // g₁(0) = start₀ + start₁
+    // g₁(1) = start₀ + start₁ + step₀ + step₁
+    // g₁(2) = start₀ + start₁ + 2step₀ + 2step₁
+    // g₁(3) = start₀ + start₁ + 3step₀ + 3step₁
+    //         (where product_sum₀ = start₀ + start₁ and product_sum₁ = product_sum₀ + step₀ + step₁)
```

Comment:
//         where products_sum₀ = start₀ + start₁ and product_sumᵢ = product_sum₀ + (i - 1) * (step₀ + step₁)